messages_view: Notify user when messages are not being marked as read.

chdinesh1089 opened the pull request:
Notifies user when messages are not being marked as read
through a banner that lets them mark all messages in
the narrow as read.

Fixes: #18768

This is another PR working on #18768 but with a different UX. see https://github.com/zulip/zulip/pull/18807#issuecomment-888544905 

(The other PR #18807 allows a bit more functionality by marking only the messages above selected message and marks the remaining messages on scrolling through them. But we wanted to try this simpler version than that.)

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->
Tested by moving into search narrow and to different narrows. Checked the button is working.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

![image](https://user-images.githubusercontent.com/44665669/127906531-f9d01102-ef38-474e-abcc-702218b5327d.png)

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

chdinesh1089 commented:
@alya, Could you review this? I'll keep this branch running at http://zulip.chdinesh1089.zulipdev.org:9991/ so it would be quicker for you to play around with this instead of having to fetch it.

(Continuing our discussion in #18807. We wanted a simpler version of the banner that would stay near the last message in the narrow)

alya commented:
Hm, for testing it, could you make there be some streams with unread
messages? I'm not seeing the banner, probably because everything is read?

On Mon, Aug 9, 2021 at 9:11 AM Dinesh Ch ***@***.***> wrote:

> @alya <https://github.com/alya>, Could you review this? I'll keep this
> branch running at http://zulip.chdinesh1089.zulipdev.org:9991/ so it
> would be quicker for you to play around with this instead of having to
> fetch it.
>
> (Continuing our discussion in #18807
> <https://github.com/zulip/zulip/pull/18807>. We wanted a simpler version
> of the banner that would stay near the last message in the narrow)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/19462#issuecomment-895210599>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP6IUTQK6SWFRPGO64L5L3T37HZXANCNFSM5BNITQ2Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

chdinesh1089 commented:
sure, sent a few messages with strings 'asdf' and 'abcd'. Searching for either of those will show the banner.

chdinesh1089 commented:
@alya ping.

alya commented:
Sorry I lost track of this one @chdinesh1089!

I just gave it another look, and the banner looks fine to me. Can we get this integrated?

Also, @amanagr FYI as another thing to keep in mind when thinking about positioning for the "Next topic" button.

timabbott commented:
I applied these changes, which are primarily:
* Moving a big note about a subtle detail from the commit message to a comment.
* Adjusting the strings
* Adding a comment about a bug in the logic, and also simplifying the calculation for what unread messages exist.

```
diff --git a/static/js/message_scroll.js b/static/js/message_scroll.js
index 5e689f5236..8da74c1bc1 100644
--- a/static/js/message_scroll.js
+++ b/static/js/message_scroll.js
@@ -202,10 +202,17 @@ export function initialize() {
                 unread_ops.notify_server_messages_read(messages, {from: "pointer"});
             } else if (
                 unread.get_unread_messages(messages).length !== 0 &&
+                // This filter check might seem redundant, but it's
+                // possible this logic, which runs after a delay, lost
+                // a race with switching to another view, like Recent
+                // Topics, and we don't want to displ[ay this banner
+                // in such a view.
+                //
+                // This can likely be fixed more cleanly with another approach.
                 narrow_state.filter() !== undefined
             ) {
-                // Ask user if they want to mark messages as read when
-                // their interactions would've marked messages as read.
+                // Display banner notifying user that this view
+                // doesn't mark messages as read.
                 $("#mark_as_read_turned_off_banner").show();
             }
         }
diff --git a/static/js/unread_ui.js b/static/js/unread_ui.js
index 7ad2dc009c..57bc11fc85 100644
--- a/static/js/unread_ui.js
+++ b/static/js/unread_ui.js
@@ -97,15 +97,14 @@ export function initialize() {
     update_unread_counts();
 
     $("#mark_as_read_turned_off_banner").html(render_mark_as_read_turned_off_banner());
-    $("#mark_all_read").on("click", () => {
-        const message_list = message_lists.current;
-
-        // Mark all messages from first unread in the view to the selected message.
-        const messages_to_mark_as_read = message_list.message_range(
-            message_list.first_unread_message_id(),
-            message_list.data.selected_id(),
-        );
-        notify_server_messages_read(messages_to_mark_as_read);
+    $("#mark_view_read").on("click", () => {
+        // Mark all messages in the current view as read.
+        //
+        // BUG: This logic only supports marking messages visible in
+        // the present view as read; we need a server API to mark
+        // every message matching the current search as read.
+        const unread_messages = message_lists.current.data.all_messages().filter((message) => unread.message_unread(message));
+        notify_server_messages_read(unread_messages);
 
         $("#mark_as_read_turned_off_banner").hide();
     });
diff --git a/static/templates/mark_as_read_turned_off_banner.hbs b/static/templates/mark_as_read_turned_off_banner.hbs
index 11593ce69a..7da4c9e0af 100644
--- a/static/templates/mark_as_read_turned_off_banner.hbs
+++ b/static/templates/mark_as_read_turned_off_banner.hbs
@@ -1,7 +1,9 @@
-<p id="mark_as_read_turned_off_content">{{t 'Messages are not being marked as read, so that it\'s easy to return to them later' }}</p>
+<p id="mark_as_read_turned_off_content">
+    {{t 'To preserve your reading state, this view does not mark messages as read.' }}
+</p>
 <div id="mark_as_read_controls">
-    <button id="mark_all_read" class="btn btn-warning" title="{{t 'Mark messages as read' }}">
-        {{t 'Mark messages as read' }}
+    <button id="mark_view_read" class="btn btn-warning" title="{{t 'Mark as read' }}">
+        {{t 'Mark as read' }}
     </button>
     <button type="button" class="mark_as_read_close close">×</button>
 </div>
```

chdinesh1089 commented:
Done. I made these slight changes to the existing commit:
* Extracted `$("#mark_as_read_turned_off_banner").show()` to `unread_ui.notify_user_messages_arent_being_marked_read();` to avoid having this comment at two places and possibility of only one being updated later 
```
                // Display banner notifying user that this view
                // doesn't mark messages as read.
```
* In addition to `narrow_state.filter() !== undefined`, added `message_lists.current === event.msg_list` also to consider the case of moving to a narrow where messages could be marked read.

timabbott commented:
This is great.  Since I think we want to only display the unread notice if the message list contains at least one unread message, I added a quick check for that condition; here's the total changes.

```
diff --git a/frontend_tests/node_tests/example7.js b/frontend_tests/node_tests/example7.js
index e8f0145c7f..a6543101bd 100644
--- a/frontend_tests/node_tests/example7.js
+++ b/frontend_tests/node_tests/example7.js
@@ -114,7 +114,7 @@ run_test("unread_ops", ({override, override_rewire}) => {
     override(message_lists.home, "show_message_as_read", () => {});
     override(notifications, "close_notification", () => {});
     override(unread_ui, "update_unread_counts", () => {});
-    override(unread_ui, "notify_user_messages_arent_being_marked_read", () => {});
+    override(unread_ui, "notify_messages_remain_unread", () => {});
 
     // Set up a way to capture the options passed in to channel.post.
     let channel_post_opts;
@@ -128,6 +128,7 @@ run_test("unread_ops", ({override, override_rewire}) => {
     // toggle it easily from within the test (and avoid complicated
     // data setup).
     override(message_lists.current, "can_mark_messages_read", () => can_mark_messages_read);
+    override(message_lists.current, "has_unread_messages", () => true);
 
     // First, test for a message list that cannot read messages.
     can_mark_messages_read = false;
diff --git a/static/js/message_scroll.js b/static/js/message_scroll.js
index 7d6c2b4719..cdc13f013e 100644
--- a/static/js/message_scroll.js
+++ b/static/js/message_scroll.js
@@ -213,7 +213,7 @@ export function initialize() {
                 narrow_state.filter() !== undefined &&
                 message_lists.current === event.msg_list
             ) {
-                unread_ui.notify_user_messages_arent_being_marked_read();
+                unread_ui.notify_messages_remain_unread();
             }
         }
     });
diff --git a/static/js/unread_ops.js b/static/js/unread_ops.js
index 8a665d2830..c2a5c2f3c6 100644
--- a/static/js/unread_ops.js
+++ b/static/js/unread_ops.js
@@ -102,8 +102,14 @@ export function notify_server_message_read(message, options) {
 export function process_scrolled_to_bottom() {
     if (message_lists.current.can_mark_messages_read()) {
         mark_current_list_as_read();
-    } else {
-        unread_ui.notify_user_messages_arent_being_marked_read();
+        return;
+    }
+
+    // For message lists that don't support marking messages as read
+    // automatically, we display a banner offering to let you mark
+    // them as read manually, only if there are unreads present.
+    if (message_lists.current.has_unread_messages()) {
+        unread_ui.notify_messages_remain_unread();
     }
 }
 
diff --git a/static/js/unread_ui.js b/static/js/unread_ui.js
index 9980804e0e..89f42ec032 100644
--- a/static/js/unread_ui.js
+++ b/static/js/unread_ui.js
@@ -93,7 +93,7 @@ export function should_display_bankruptcy_banner() {
     return false;
 }
 
-export function notify_user_messages_arent_being_marked_read() {
+export function notify_messages_remain_unread() {
     $("#mark_as_read_turned_off_banner").show();
 }
 
```

Pushed back here and deploying to chat.zulip.org for final feedback, but this should be good to merge aside from new design changes.

andersk commented:
This banner is incorrectly showing up in the All messages view if it’s the view the page loads with.

timabbott commented:
That All messages issue was fixed by adding this line to the `initialize` method for `unread_ui.js`; looks like we forgot to update the thread.
```
+    $("#mark_as_read_turned_off_banner").hide();
```


Merging this, thanks @chdinesh1089 for making this happen!
